### PR TITLE
fix(instantsearch.js): fix user token not being set in time for the first query

### PR DESCRIPTION
**Pull request by shaejaz**
[FX-2955]

**Summary**
Double queries happen when the state of instantsearch and insights no longer matches at the time of the first query on page load.

- The `search-insights` `onUserTokenChange` and `onAuthenticatedUserTokenChange` callbacks are not able to set the `user-token` by the time the first search query is sent. When the `user-token` is changed by these callbacks, the state has changed from the first query. If a new query is scheduled after this(for example with dynamic widgets), the cache misses due to the changed state and second query is made.

- When `useCookie` is set to true, insights needs to create a anonymous token and save it as a cookie. Instantsearch would have the token updated with the callback from before, but this does not happen in time for the first query. So anonymous token generation and saving in cookies is now also handled by instantsearch.

- The insights queue is now being access in the middleware's `started` function to make sure any token set before instantsearch was started are part of the state by the time the first query is made.

[FX-2955]: https://algolia.atlassian.net/browse/FX-2955?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts [hunk 1]
```diff
@@ -192,11 +199,29 @@ export function createInsightsMiddleware<
           instantSearchInstance.emit('error', new Error(errorMessage));
         }
       },
+      // eslint-disable-next-line complexity
```
Comment (Haroenv) [inline packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts:202]: what is the complexity about? can it be avoided?

File: packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts [hunk 2]
```diff
@@ -192,11 +199,29 @@ export function createInsightsMiddleware<
...
         insightsClient('addAlgoliaAgent', 'insights-middleware');
 
         helper = instantSearchInstance.mainHelper!;
 
+        const { queue: queueAtStart } = insightsClient;
+
+        if (Array.isArray(queueAtStart)) {
+          [queuedUserToken, queuedAuthenticatedUserToken] = [
+            'setUserToken',
+            'setAuthenticatedUserToken',
```
Comment (Haroenv) [inline packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts:213]: no init here, maybe we should get the value again?

File: packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts [hunk 3]
```diff
@@ -287,12 +369,30 @@ export function createInsightsMiddleware<
...
+        } else if (anonymousUserToken) {
+          setUserToken(anonymousUserToken, anonymousUserToken, undefined);
+
+          if (insightsInitParams?.useCookie || queuedInitParams?.useCookie) {
+            const d = new Date();
+            d.setTime(
+              d.getTime() +
+                (insightsInitParams?.cookieDuration ||
+                  queuedInitParams?.cookieDuration ||
+                  30 * 24 * 60 * 60 * 1000 * 6)
```
Comment (Haroenv) [inline packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts:381]: maybe a comment that says "six months"?

File: packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts [hunk 4]
```diff
@@ -287,12 +369,30 @@ export function createInsightsMiddleware<
             queuedUserToken,
             queuedAuthenticatedUserToken
           );
+        } else if (anonymousUserToken) {
+          setUserToken(anonymousUserToken, anonymousUserToken, undefined);
```
Comment (Haroenv) [inline packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts:373]: would like to see the setUserToken function simplified (can be separate pr) to not have three arguments

**Review by Haroenv — COMMENTED** [review]
looks good! probably just need added tests

File: packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts (hunk 1 above)
Comment (shaejaz) [reply packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts:202]: I'll try to extract some functionality into separate functions, should fix this.

File: packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts (hunk 3 above)
Comment (shaejaz) [reply packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts:381]: Oop, yeah this part needs to be cleaned

File: packages/instantsearch.js/src/middlewares/__tests__/createInsightsMiddleware.ts [hunk 5]
```diff
@@ -437,7 +437,7 @@ describe('insights', () => {
       );
     });
 
-    it('warns when userToken is not set', () => {
```
Comment (shaejaz) [inline packages/instantsearch.js/src/middlewares/__tests__/createInsightsMiddleware.ts:440]: Not sure if this is needed now, since an anonymous token would always be set for queries.

File: packages/instantsearch.js/src/lib/__tests__/InstantSearch-integration-test.ts [hunk 6]
```diff
@@ -489,6 +489,7 @@ describe('network requests', () => {
             "params": {
               "clickAnalytics": true,
               "query": "",
+              "userToken": "cookie-key",
```
Comment (shaejaz) [inline packages/instantsearch.js/src/lib/__tests__/InstantSearch-integration-test.ts:492]: if a cookie is present, it will always be used if the token is not no other way

Comment (Haroenv) [reply packages/instantsearch.js/src/lib/__tests__/InstantSearch-integration-test.ts:492]: can you clarify what you mean? not sure I understand your comment

File: packages/instantsearch.js/src/middlewares/__tests__/createInsightsMiddleware.ts (hunk 5 above)
Comment (Haroenv) [reply packages/instantsearch.js/src/middlewares/__tests__/createInsightsMiddleware.ts:440]: yes, there's no realistic way to get this warning anymore. Still fine to keep at as a worst-case warning but I don't see how you could _not_ have a token anymore with these changes.

File: packages/instantsearch.js/src/middlewares/__tests__/createInsightsMiddleware.ts [hunk 7]
```diff
@@ -777,38 +777,20 @@ See documentation: https://www.algolia.com/doc/guides/building-search-ui/going-f
...
 
-      instantSearchInstance.start();
-
-      await wait(0);
-
-      expect(
-        (instantSearchInstance.mainHelper!.state as PlainSearchParameters)
-          .userToken
-      ).toEqual('def');
+      expect(getUserToken()).toEqual('def');
```
Comment (Haroenv) [inline packages/instantsearch.js/src/middlewares/__tests__/createInsightsMiddleware.ts:793]: we also want to test the token of the search parameters is `def` here

File: packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts [hunk 8]
```diff
@@ -269,13 +306,58 @@ export function createInsightsMiddleware<
...
+        }
+
+        let authenticatedUserTokenFromInit: string | undefined;
+        let userTokenFromInit: string | undefined;
+
+        // By the time the first query is sent, the token would not be set by the insights
+        // onChange callbacks. It is explicitly being set here so that the first query
+        // has the initial tokens set and inturn a second query isn't automatically made
+        // when the onChange callback actually changes the state.
+        if (insightsInitParams) {
```
Comment (Haroenv) [inline packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts:330]:
```suggestion
        // When the first query is sent, the token is possibly not yet be set by
        // the insights onChange callbacks (if insights isn't yet loaded).
        // It is explicitly being set here so that the first query has the
        // initial tokens set and ensure a second query isn't automatically
        // made when the onChange callback actually changes the state.
        if (insightsInitParams) {
```

File: tests/common/shared/insights.ts [hunk 9]
```diff
@@ -76,7 +76,7 @@ export function createInsightsTests(
         });
 
         // initial calls because the middleware is attached
-        expect(window.aa).toHaveBeenCalledTimes(6);
+        expect(window.aa).toHaveBeenCalledTimes(7);
```
Comment (Haroenv) [inline tests/common/shared/insights.ts:79]: the change here is because of the automatic token being set?

File: packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts [hunk 10]
```diff
@@ -269,13 +306,58 @@ export function createInsightsMiddleware<
...
+
         // We consider the `userToken` or `authenticatedUserToken` before an
-        // `init` call of higher importance than one from the queue.
+        // `init` call of higher importance than one from the queue and ones set
+        // from the init props to be higher than that.
+        const tokenFromInit =
+          authenticatedUserTokenFromInit || userTokenFromInit;
         const tokenBeforeInit =
           authenticatedUserTokenBeforeInit || userTokenBeforeInit;
         const queuedToken = queuedAuthenticatedUserToken || queuedUserToken;
```
Comment (Haroenv) [inline packages/instantsearch.js/src/middlewares/createInsightsMiddleware.ts:346]: this should be `tokenFromQueue`

File: packages/instantsearch.js/src/middlewares/__tests__/createInsightsMiddleware.ts (hunk 5 above)
Comment (Haroenv) [reply packages/instantsearch.js/src/middlewares/__tests__/createInsightsMiddleware.ts:440]: the test should be removed, not skipped :)

**Review by Haroenv — COMMENTED** [review]
lgtm when the skipped test is removed

**Review by Haroenv — APPROVED** [review]

**Review by dhayab — APPROVED** [review]
Kudos for solving this 👏
